Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
-        cur->next = prev;
+    void AddOperation(NSOperation* operation, unsigned int priority, NSOperationQueue* queue, SEL selector) {
+        {
+            std::lock_guard<std::mutex> lock(_lock);
 
-        if (next) {
-            prev = cur;
-            cur = next;
+            StrongId<NSOperation> strongOperation = operation;
+            _operationsHoldingArea[priority].emplace_back(strongOperation);
+            if (_thread == nil || _thread.get().finished) {
+                _thread.attach([[NSThread alloc] initWithTarget:queue selector:selector object:nil]);
+                [_thread start];
+            }
         }
-    } while (next);
 
-    *listPtr = cur;
-}
+        _workAvailableCondition.notify_one();
+    }
 
-NSObject* NSAtomicListPop(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
+    void SignalAllWorkDone() {
+        _allWorkDoneCondition.notify_one();
     }
 
-    *listPtr = node->next;
+    void WaitUntilAllOperationsAreFinished() {
+        std::unique_lock<std::mutex> lock(_lock);
 
-    NSObject* elt = [[node->elt retain] autorelease];
-    delete node;
-    return elt;
-}
+        while (_HasMoreWorkNoLock()) {
+            _allWorkDoneCondition.wait(lock);
+        }
+    }
 
-NSObject* NSAtomicListPeek(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
+    void SetMaxConcurrentOperationCount(NSInteger count) {
+        std::lock_guard<std::mutex> lock(_lock);
+        _maxConcurrentOperationCount = count;
     }
 
-    return node->elt;
-}
+    NSInteger MaxConcurrentOperationCount() {
+        std::lock_guard<std::mutex> lock(_lock);
+        return _maxConcurrentOperationCount;
+    }
 
-void NSAtomicListInsert(NSAtomicListRef* listPtr, NSObject* elt) {
-    struct NSAtomicListNode* node = new NSAtomicListNode();
-    node->elt = elt;
+    NSArray<NSOperation*>* GetOperations() {
+        std::lock_guard<std::mutex> lock(_lock);
 
-    do {
-        node->next = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, node, node->next) != node->next);
-}
+        id ret = [NSMutableArray array];
 
-static id _mainQueue;
+        if (_curOperation != nil) {
+            [ret addObject:_curOperation];
+        }
 
-@interface NSOperationQueue () {
-    struct NSOperationQueuePriv* priv;
-    dispatch_queue_t _completionQueue;
-}
-@end
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            if (!_operationsHoldingArea[i].empty()) {
+                for (const auto& op : _operationsHoldingArea[i]) {
+                    [ret addObject:op];
+                }
+            }
 
-@implementation NSOperationQueue
-static void ClearList(NSAtomicListRef* listPtr) {
-    for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-        while (NSAtomicListPop(&listPtr[i])) {
-            ;
+            if (!_candidateOperationsToRunNext[i].empty()) {
```

Comment:
nit: Unnecessary? It should be just as many checks as iterating over an empty list.